### PR TITLE
Rewrite AGG(IF()) to AGG() FILTER()

### PR DESCRIPTION
**Pull request by yuanzhanhku**
Add an rule to rewrite
 - AGG(IF(condition, expr))
to
 - AGG(expr) FILTER (WHERE condition).

The latter plan is more efficient because:
 - the filter can be pushed down to the scan node
 - the rows not matching the condition are not aggregated
 - the IF() expression wrapper is removed.

Test plan
Added unit tests for the rule in TestRewriteAggIfToAggFilter.java. 
Added query plan tests in TestFilteredAggregations.java and covered all existing test cases for AGG() FILTER.

```
== RELEASE NOTES ==

General Changes
* Introduce a new config ``optimizer.aggregation-if-to-filter-rewrite-enabled`` and its corresponding session property ``aggregation_if_to_filter_rewrite_enabled`` to enable or disable an optimizer rule to improve the query performance of ``IF`` expressions inside aggregation functions.
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 1]
```diff
@@ -199,6 +199,7 @@
     public static final String VERBOSE_EXCEEDED_MEMORY_LIMIT_ERRORS_ENABLED = "verbose_exceeded_memory_limit_errors_enabled";
     public static final String MATERIALIZED_VIEW_DATA_CONSISTENCY_ENABLED = "materialized_view_data_consistency_enabled";
     public static final String QUERY_OPTIMIZATION_WITH_MATERIALIZED_VIEW_ENABLED = "query_optimization_with_materialized_view_enabled";
+    public static final String AGG_IF_TO_FILTER_REWRITE_ENABLED = "agg_if_to_filter_rewrite_enabled";
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:202]: AGGREGATION_IF_TO_FILTER_REWRITE_ENABLED; let's spell out "aggregation" in this PR

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 2]
```diff
@@ -1066,7 +1067,12 @@ public SystemSessionProperties(
                         QUERY_OPTIMIZATION_WITH_MATERIALIZED_VIEW_ENABLED,
                         "Enable query optimization with materialized view",
                         featuresConfig.isQueryOptimizationWithMaterializedViewEnabled(),
-                        true));
+                        true),
+                booleanProperty(
+                        AGG_IF_TO_FILTER_REWRITE_ENABLED,
+                        "Enable rewriting AGG(IF()) to AGG() FILTER",
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1073]: Enable rewriting the if expression inside an aggregation function to a filter clause outside aggregation.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java [hunk 3]
```diff
@@ -1786,4 +1787,17 @@ public FeaturesConfig setQueryOptimizationWithMaterializedViewEnabled(boolean va
...
         return this;
     }
+
+    public boolean isAggIfToFilterRewriteEnabled()
+    {
+        return aggIfToFilterRewriteEnabled;
+    }
+
+    @Config("agg-if-to-filter-rewrite-enabled")
+    @ConfigDescription("Enable rewriting AGG(IF()) to AGG() FILTER")
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java:1797]: Same; update description and "agg"

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java [hunk 4]
```diff
@@ -1786,4 +1787,17 @@ public FeaturesConfig setQueryOptimizationWithMaterializedViewEnabled(boolean va
         this.queryOptimizationWithMaterializedViewEnabled = value;
         return this;
     }
+
+    public boolean isAggIfToFilterRewriteEnabled()
+    {
+        return aggIfToFilterRewriteEnabled;
+    }
+
+    @Config("agg-if-to-filter-rewrite-enabled")
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java:1796]: we usually add a prefix "optimizer." to optimizer related config changes; check other examples in this class

**Review by highker — COMMENTED** [review]
still reviewing; but quick nit: let's change all "agg" into "aggregation" in this PR...

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                                predicate),
+                        aggregations.build(),
+                        aggregationNode.getGroupingSets(),
+                        aggregationNode.getPreGroupedVariables(),
+                        aggregationNode.getStep(),
+                        aggregationNode.getHashVariable(),
+                        aggregationNode.getGroupIdVariable()));
+    }
+
+    private boolean ShouldRewriteAggregation(AggregationNode.Aggregation aggregation, ProjectNode sourceProject)
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java:172]: nit: `shouldRewriteAggregation`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        }
+        if (aggregation.getFilter().isPresent() || aggregation.getMask().isPresent()) {
+            // Do not rewrite the aggregation if it already has a filter or mask.
+            return false;
+        }
+        RowExpression sourceExpression = sourceProject.getAssignments().get((VariableReferenceExpression) aggregation.getArguments().get(0));
+        if (!(sourceExpression instanceof SpecialFormExpression)) {
+            return false;
+        }
+        SpecialFormExpression expr = (SpecialFormExpression) sourceExpression;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java:186]: spell out "expression"

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        // Only rewrite the aggregation if the else branch is not present.
+        return expr.getForm() == SpecialFormExpression.Form.IF && Expressions.isNull(expr.getArguments().get(2));
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java:188]: static import `IF`
```java
expr.getForm() == IF

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java (hunk 5 above)
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java:172]: static import `Aggregation`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregationNode, Captures captures, Context context)
+    {
+        ProjectNode sourceProject = captures.get(CHILD);
+
+        Set<AggregationNode.Aggregation> aggregationsToRewrite = aggregationNode.getAggregations().values().stream()
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java:84]: static import `Aggregation`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+                .filter(aggregation -> ShouldRewriteAggregation(aggregation, sourceProject))
+                .collect(toSet());
+        if (aggregationsToRewrite.isEmpty()) {
+            return Result.empty();
+        }
+
+        // Extract the VariableReferenceExpressions from these aggregations.
+        // Multiple aggregations may reference the same input. We use a set to dedup them so that we only do the rewrite once per input IF expression.
+        Set<VariableReferenceExpression> aggregationReferences =
+                aggregationsToRewrite.stream().map(a -> (VariableReferenceExpression) a.getArguments().get(0)).collect(toSet());
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java:94]: - s/a/aggregation
- s/toSet/toImmutableSet

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+        // Get the corresponding assignments in the input project.
+        Map<VariableReferenceExpression, RowExpression> sourceAssignments = sourceProject.getAssignments().entrySet().stream()
+                .filter(entry -> aggregationReferences.contains(entry.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java:99]: s/toMap/toImmutableMap

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+        Assignments.Builder newAssignments = Assignments.builder();
+        // We don't remove the IF expression now in case the aggregation has other references to it. These will be cleaned up by other pruning rules.
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java:102]: I guess it's PruneUnreferencedOutputs? Maybe let's highlight the rule here

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+        newAssignments.putAll(sourceProject.getAssignments());
+
+        // Map from the aggregation reference to the IF condition reference.
+        Map<VariableReferenceExpression, VariableReferenceExpression> aggregationReferenceToConditionReference = new HashMap<>();
+        // Map from the aggregation reference to the IF result reference.
+        Map<VariableReferenceExpression, VariableReferenceExpression> aggregationReferenceToIfResultReference = new HashMap<>();
+
+        for (Map.Entry<VariableReferenceExpression, RowExpression> entry : sourceAssignments.entrySet()) {
+            VariableReferenceExpression outputVariable = entry.getKey();
+            SpecialFormExpression ifExpr = (SpecialFormExpression) entry.getValue();
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java:112]: - spell out ifExpression
- when we build `sourceAssignments`, we didn't filter on IF-only expressions. So there could be other types of expressions here. We probably wanna reuse some code in `ShouldRewriteAggregation`. Worth adding a test to cover this.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+
+            RowExpression condition = ifExpr.getArguments().get(0);
+            VariableReferenceExpression conditionReference = context.getVariableAllocator().newVariable(condition);
+            newAssignments.put(conditionReference, condition);
+            aggregationReferenceToConditionReference.put(outputVariable, conditionReference);
+
+            RowExpression ifResult = ifExpr.getArguments().get(1);
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java:119]: - s/ifResult/trueResult
- also, we didn't validate if the if expr has 2 or 3 arguments, which we don't support the latter one.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+                aggregations.put(output, aggregation);
+                continue;
+            }
+            VariableReferenceExpression aggregationReference = (VariableReferenceExpression) aggregation.getArguments().get(0);
+            CallExpression callExpression = aggregation.getCall();
+            VariableReferenceExpression mask = aggregationReferenceToConditionReference.get(aggregationReference);
+            aggregations.put(output, new AggregationNode.Aggregation(
+                    new CallExpression(callExpression.getDisplayName(),
+                            callExpression.getFunctionHandle(), callExpression.getType(), ImmutableList.of(
+                            aggregationReferenceToIfResultReference.get(aggregationReference))),
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java:141]: one param per line
```java
                    new CallExpression(
                            callExpression.getDisplayName(),
                            callExpression.getFunctionHandle(),
                            callExpression.getType(),
                            ImmutableList.of(aggregationReferenceToIfResultReference.get(aggregationReference))),

**Review by highker — COMMENTED** [review]
more comments on logic

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java (hunk 12 above)
Comment (yuanzhanhku) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java:112]: sourceAssignments only has the references in the aggregations filtered by shouldRewriteAggregation. So it only has IF expressions. testFireForOneOfTwoAggregations covers this case.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java (hunk 13 above)
Comment (yuanzhanhku) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java:119]: shouldRewriteAggregation already validated that the If expression doesn't have else results. So we are good here.

**Comment by yuanzhanhku** [comment]
The order of the aggregations in the Aggregation node seems not deterministic, so I made a few changes to generate the new expressions in the order of the VariableReferenceExpression names.

**Comment by highker** [comment]
Quick nit: could you add the "release note" section to this github page as well? Check existing merged PRs with release note as examples. This should be "general changes" with something like "introduce a new config and session property to blah blah blah". Note that a release note should be extremely user-facing, mean that we could use sentences like "optimize if expressions in aggregation functions to improve performance".

File: presto-expressions/src/main/java/com/facebook/presto/expressions/LogicalRowExpressions.java [hunk 15]
```diff
@@ -132,12 +132,12 @@ public static RowExpression or(RowExpression... expressions)
...
     }
 
-    public static RowExpression or(Collection<RowExpression> expressions)
+    public static RowExpression or(Collection<? extends RowExpression> expressions)
     {
         return binaryExpression(OR, expressions);
     }
 
-    public static RowExpression binaryExpression(Form form, Collection<RowExpression> expressions)
+    public static RowExpression binaryExpression(Form form, Collection<? extends RowExpression> expressions)
```
Comment (highker) [inline presto-expressions/src/main/java/com/facebook/presto/expressions/LogicalRowExpressions.java:140]: It would be good to separate out to another commit before this. Many other helpers (e.g., `and`) can be changed as well.

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java [hunk 16]
```diff
@@ -126,28 +179,42 @@ public void rewriteAddFilterWithMultipleFilters()
...
                                                 "custkey", "custkey")))));
+
+        assertPlan(
+                "SELECT sum(IF(totalprice > 0, totalprice)), sum(IF(custkey > 0, custkey)) FROM orders",
+                anyTree(
+                        filter(
+                                "(\"totalprice\" > 0E0 OR \"custkey\" > BIGINT '0')",
+                                tableScan(
+                                        "orders", ImmutableMap.of("totalprice", "totalprice",
+                                                "custkey", "custkey")))));
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java:190]: make sure for each function:
- either we put all params into one line or
- make one param per line and leave the first line empty

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+        }
+
+        // Get the corresponding assignments in the input project.
+        // The aggregationReferences only has the aggregations to rewrite, thus the sourceAssignments only has IF expressions with NULL false results.
+        // Multiple aggregations may reference the same input. We use a map to dedup them based on the VariableReferenceExpression, so that we only do the rewrite once per input
+        // IF expression.
+        // The order of sourceAssignments determines the order of generating the new variables for the IF conditions and results. We use a sorted map to get a deterministic
+        // order based on the name of the VariableReferenceExpressions.
+        Map<VariableReferenceExpression, RowExpression> sourceAssignments = aggregationsToRewrite.stream()
+                .map(a -> (VariableReferenceExpression) a.getArguments().get(0))
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java:102]: s/a/aggregation

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+                .collect(toImmutableSortedMap(VariableReferenceExpression::compareTo, identity(), v -> sourceProject.getAssignments().get(v), (v1, v2) -> v1));
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java:103]: nit: break the params of `toImmutableSortedMap` to 4 lines for readability:
- replace `v` with `variable`
- replace `v1` and `v2` with `left` and `right`

**Review by highker — APPROVED** [review]

File: presto-expressions/src/main/java/com/facebook/presto/expressions/LogicalRowExpressions.java (hunk 15 above)
Comment (yuanzhanhku) [reply presto-expressions/src/main/java/com/facebook/presto/expressions/LogicalRowExpressions.java:140]: Done.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+            // Do not rewrite the aggregation if it already has a filter or mask.
+            return false;
+        }
+        RowExpression sourceExpression = sourceProject.getAssignments().get((VariableReferenceExpression) aggregation.getArguments().get(0));
+        if (!(sourceExpression instanceof SpecialFormExpression)) {
+            return false;
+        }
+        SpecialFormExpression expression = (SpecialFormExpression) sourceExpression;
+        // Only rewrite the aggregation if the else branch is not present.
+        return expression.getForm() == IF && Expressions.isNull(expression.getArguments().get(2));
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java:195]: You could also add when the else part is the null literal like IF(x, y, null)

**Review by kaikalur — COMMENTED** [review]
So SET_AGG works differently for these two cases:
```

presto:di> select set_agg(if(x=1,y)) from (select 1 x, 2 y union all select null x, 20 y union all select 1 x, null y) group by y;
 _col0
--------
 [2]
 [null]
 [null]
(3 rows)

Query 20210805_003013_03168_fbgd3, FINISHED, 195 nodes
Splits: 3,139 total, 3,139 done (100.00%)
0:02 [0 rows, 0B] [0 rows/s, 0B/s]

presto:di> select set_agg( y) filter (where x=1) from (select 1 x, 2 y union all select null x, 20 y union all select 1 x, null y) group by y;
 _col0
--------
 NULL
 [null]
 [2]
 ```
 
 So you may want to see if there are aggregation properties needed for this to work properly.

**Comment by yuanzhanhku** [comment]
> So SET_AGG works differently for these two cases:
> 
> ```
> 
> presto:di> select set_agg(if(x=1,y)) from (select 1 x, 2 y union all select null x, 20 y union all select 1 x, null y) group by y;
>  _col0
> --------
>  [2]
>  [null]
>  [null]
> (3 rows)
> 
> Query 20210805_003013_03168_fbgd3, FINISHED, 195 nodes
> Splits: 3,139 total, 3,139 done (100.00%)
> 0:02 [0 rows, 0B] [0 rows/s, 0B/s]
> 
> presto:di> select set_agg( y) filter (where x=1) from (select 1 x, 2 y union all select null x, 20 y union all select 1 x, null y) group by y;
>  _col0
> --------
>  NULL
>  [null]
>  [2]
> ```
> 
> So you may want to see if there are aggregation properties needed for this to work properly.

Thanks for catching this. I will make a fix and add the tests.

**Comment by kaikalur** [comment]
In fact, I think we should do this only for numerical aggs (or an allowlist to start with SUM/COUNT/MIN/MAX)

**Comment by yuanzhanhku** [comment]
> In fact, I think we should do this only for numerical aggs (or an allowlist to start with SUM/COUNT/MIN/MAX)

Right. I am planning only do this rewrite for numerical aggs. In addition to SUM/COUNT/MIN/MAX, I see there are many use cases for approx_distinct/ variance/approx_percentile as well. It would be good to enable the rewrite for these too.

**Comment by kaikalur** [comment]
Actually, there is more:
* Make sure the NULL behavior of the agg is "not called on null"
* Make sure the condition, then part of the if expression are both deterministic

Also, when you are at it, keep applying the simplification logic iteratively so things like IF(p1, IF(p2, x)) are also handled (I have seen those in tool generated code). 

Use the RowExpressionInterpreter to simplify the IF expression before checking if the else part is null. It does some more const expr eval so good to be comprehensive.

**Comment by kaikalur** [comment]
Just to be clear, if you do those things, there should be no need to special csase.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+                    aggregation.getOrderBy(),
+                    aggregation.isDistinct(),
+                    Optional.of(aggregationReferenceToConditionReference.get(aggregationReference))));
+            masks.add(mask);
+        }
+
+        RowExpression predicate = TRUE_CONSTANT;
+        if (!aggregationNode.hasNonEmptyGroupingSet() && aggregationsToRewrite.size() == aggregationNode.getAggregations().size()) {
+            // All aggregations are rewritten by this rule. We can add a filter with all the masks to make the query more efficient.
+            predicate = or(masks.build());
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java:159]: This or might actually cause more slowdown than help because all masks have to be evaluated for every row anyway. I don't think this helps.

**Comment by yuanzhanhku** [comment]
> Actually, there is more:
> 
> * Make sure the NULL behavior of the agg is "not called on null"
Great point on checking isCalledOnNullInput(). It is indeed the root cause of the behavior change as this rewrite filters out the NULL values.
> * Make sure the condition, then part of the if expression are both deterministic

Could you please elaborate why non-deterministic functions matter in this case? Do you have some examples when this could cause issues?
> 
> Also, when you are at it, keep applying the simplification logic iteratively so things like IF(p1, IF(p2, x)) are also handled (I have seen those in tool generated code).
Yes, this is a nice additional optimization we can do. Basically, we can rewrite AGG(IF(p1, IF(p2, x))) to AGG(x) FILTER(WHERE p1 AND p2).
Actually, it might be better to have a separate rewrite to inline the IF expressions in this case. i.e., rewrite IF(p1, IF(p2, x)) to IF(p1 AND p2, x). This rewrite can be applied for non-agg functions as well. It might improve performance because it simplifies the function?
> 
> Use the RowExpressionInterpreter to simplify the IF expression before checking if the else part is null. It does some more const expr eval so good to be comprehensive.
The SimplifyRowExpressions rules simplifies all expressions. Might be better to avoid doing the duplicate optimization here?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java (hunk 20 above)
Comment (yuanzhanhku) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java:159]: I think it helps because
* The filter could be pushed down to the scan node which might be able to evaluated very efficiently, e.g., if it is on some partition columns, the partition columns are using dictionary encoding. This only need to evaluated once per dictionary item. 
* The filter can be used to prune the partitions/splits based on column stats.

Besides, the AGG() FILTER implementation also adds this predicate. It is better to keep the same behavior:
https://github.com/prestodb/presto/blob/04138bc26a4fc801630fe936463d89d071d00c0a/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementFilteredAggregations.java#L121

**Comment by kaikalur** [comment]
> > Actually, there is more:
> > 
> > * Make sure the NULL behavior of the agg is "not called on null"
> 
> Great point on checking isCalledOnNullInput(). It is indeed the root cause of the behavior change as this rewrite filters out the NULL values.

> 
> > * Make sure the condition, then part of the if expression are both deterministic
> 
> Could you please elaborate why non-deterministic functions matter in this case? Do you have some examples when this could cause issues?

It's one of those things for example random() is called once per row I think but if you put in a then part, that might work differently. Not sure. So let's be conservative (for corner cases).

> 
> > Also, when you are at it, keep applying the simplification logic iteratively so things like IF(p1, IF(p2, x)) are also handled (I have seen those in tool generated code).
> 
> Yes, this is a nice additional optimization we can do. Basically, we can rewrite AGG(IF(p1, IF(p2, x))) to AGG(x) FILTER(WHERE p1 AND p2).
> Actually, it might be better to have a separate rewrite to inline the IF expressions in this case. i.e., rewrite IF(p1, IF(p2, x)) to IF(p1 AND p2, x). This rewrite can be applied for non-agg functions as well. It might improve performance because it simplifies the function?

I thought about it. Sure you can do that in a separate PR

> 
> > Use the RowExpressionInterpreter to simplify the IF expression before checking if the else part is null. It does some more const expr eval so good to be comprehensive.
> 
> The SimplifyRowExpressions rules simplifies all expressions. Might be better to avoid doing the duplicate optimization here?

That's fine. It's good to make sure it's simpified to get max benefit.

**Comment by yuanzhanhku** [comment]
Thanks Sreeni for all the suggestions. Made the changes in https://github.com/prestodb/presto/pull/16566.
